src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.using System;
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Sockets.Internal.Formatters
+{
+    public  class ServerSentEventsMessageParser
+    {
+
+        public static ParsePhase ParseMessage(ReadableBuffer buffer, out ReadCursor consumed, out ReadCursor examined, out Message message)
+        {
+            const byte ByteCR = (byte)'\r';
+            const byte ByteLF = (byte)'\n';
+            const byte ByteSpace = (byte)' ';
+            bool firstLine = true;
+            byte[] data = null;
+            consumed = buffer.Start;
+            examined = buffer.End;
+            message = new Message();
+            MessageType messageType = MessageType.Text;
+            var reader = new ReadableBufferReader(buffer);
+
+            var start = consumed;
+            var end = examined;
+
+            while (!reader.End)
+            {
+                var span = reader.Span;
+                var backup = reader;
+                var ch1 = reader.Take();
+                var ch2 = reader.Take();
```

Comment:
can you do that if there was just one byte in the buffer (which was consumed by the first `.Take()`?